Sources/TensorFlow/Optimizers/LearningRates.swift

```diff
@@ -0,0 +1,348 @@
+    ///   - learningRate: Learning rate value to transform according to this schedule.
+    func callAsFunction(step: UInt64, learningRate: Scalar) -> Scalar
+}
+
+extension LearningRateSchedule {
+    /// Returns a new learning rate schedule which is defined as the composition of this schedule
+    /// with the provided one. More specifically, the new schedule will transform the provided
+    /// learning rates as follows:
+    /// `self(step: step, learningRate: other(step: step, learningRate: learningRate))`.
+    public func composed<Schedule: LearningRateSchedule>(
+        with other: Schedule
+    ) -> ComposedLearningRateSchedule<Self, Schedule> {
+        ComposedLearningRateSchedule(schedule1: self, schedule2: other)
+    }
+}
+
+/// Composed learning rate schedule.
+///
+/// This schedule transforms the provided learning rates as follows:
+/// ```
+/// schedule1(step: step, learningRate: schedule2(step: step, learningRate: learningRate))
+/// ```
+public struct ComposedLearningRateSchedule<
+    Schedule1: LearningRateSchedule,
+    Schedule2: LearningRateSchedule
+>: LearningRateSchedule where Schedule1.Scalar == Schedule2.Scalar {
+    public typealias Scalar = Schedule1.Scalar
+
+    public let schedule1: Schedule1
+    public let schedule2: Schedule2
+
+    public func callAsFunction(step: UInt64, learningRate: Scalar) -> Scalar {
+        schedule1(step: step, learningRate: schedule2(step: step, learningRate: learningRate))
+    }
+}
+
+/// Dummy learning rate schedule that represents no schedule being used. This is useful as a
+/// default value whenever a learning rate schedule argument is used.
+public struct FixedLearningRate<Scalar: FloatingPoint>: LearningRateSchedule {
+    @inlinable
+    public func callAsFunction(step: UInt64, learningRate: Scalar) -> Scalar {
+        learningRate
+    }
+}
+
+/// Linear learning rate decay schedule.
+///
+/// The decayed learning rate is computed as follows:
+/// ```
+/// decayed = learningRate + step * slope
+/// decayedLearningRate = max(lowerBound * learningRate, decayed)
+/// ```
+public struct LinearLearningRateDecay<Scalar: FloatingPoint>: LearningRateSchedule {
+    public let slope: Scalar
+    public let lowerBound: Scalar
+    public let startStep: UInt64
+
+    /// Creates a new linear learning rate decay schedule.
+    ///
+    /// - Parameters:
+    ///   - slope: Slope of the linear decay.
+    ///   - lowerBound: Minimum decayed learning rate value as a fraction of the original learning
+    ///     rate value.
+    ///   - startStep: Step after which to start decaying the learning rate.
+    @inlinable
+    public init(slope: Scalar, lowerBound: Scalar = Scalar(0), startStep: UInt64 = 0) {
+        self.slope = slope
+        self.lowerBound = lowerBound
+        self.startStep = startStep
+    }
+
+    @inlinable
+    public func callAsFunction(step: UInt64, learningRate: Scalar) -> Scalar {
+        if step < startStep { return learningRate }
```

Comment:
I don't know what the particular coding style of the developers of TensorFlow for Swift have, but my personal preference and so I'm not really suggesting a change, but just pointing out, that a guard condition for things like this would be my preference. 

```
guard step >= startStep else {
    return learningRate
}
```

Reply:
Yes, that's a good question. I'm not sure what's the most common convention. Maybe @rxwei can comment on the style conventions for this library.